alf/bin/grid_search.py

```diff
@@ -125,21 +174,34 @@ def run(self):
 
         process_pool = multiprocessing.Pool(
             processes=max_worker_num, maxtasksperchild=1)
+        device_queue = self._init_device_queue(max_worker_num)
 
         task_count = 0
         for values in itertools.product(*param_values):
             parameters = dict(zip(param_keys, values))
             root_dir = "%s/%d" % (FLAGS.root_dir, task_count)
             process_pool.apply_async(
                 func=self._worker,
-                args=[root_dir, parameters],
+                args=[root_dir, parameters, device_queue],
```

Comment:
it is simpler to directly pass device_id or job_id instead of device_queue here.

Reply:
The problem with that is that the total number of jobs might be bigger than number of workers. So the scheduling is actually dynamic instead of static. I only know which device is available after one worker finishes and put the device back into the queue.